### ISSUE TITLE
Problem about package dependencies when building from source

### ISSUE THREAD

**Issue by IamJasonYe** [SMRT-AIST/fast_gicp#28]
I have a problem building the package in my catkin workspace. It has a bunch of error information about the eigen3 library. So I would like to know which version code base of eigen3 are you using to compile the package? My current eigen3 library is the newest one built from source.

Here is a part of the build error information:
/home/jason/Workspace/catkin_ws/src/fast_gicp/src/fast_gicp/gicp/fast_gicp.cpp:5:27:   required from here
/home/jason/Workspace/catkin_ws/src/fast_gicp/include/fast_gicp/gicp/impl/fast_gicp_impl.hpp:163:30: error: call to ‘double fast_gicp::FastGICP<PointSource, PointTarget>::compute_error(const Isometry3d&, Eigen::Matrix<double, 6, 6>*, Eigen::Matrix<double, 6, 1>*) const [with PointSource = pcl::PointXYZI; PointTarget = pcl::PointXYZI; Eigen::Isometry3d = Eigen::Transform<double, 3, 1>]’ uses the default argument for parameter 2, which is not yet defined
/home/jason/Workspace/catkin_ws/src/fast_gicp/include/fast_gicp/gicp/impl/fast_gicp_impl.hpp:163:30: error: call to ‘double fast_gicp::FastGICP<PointSource, PointTarget>::compute_error(const Isometry3d&, Eigen::Matrix<double, 6, 6>*, Eigen::Matrix<double, 6, 1>*) const [with PointSource = pcl::PointXYZI; PointTarget = pcl::PointXYZI; Eigen::Isometry3d = Eigen::Transform<double, 3, 1>]’ uses the default argument for parameter 3, which is not yet defined
make[2]: *** [CMakeFiles/fast_gicp.dir/src/fast_gicp/gicp/fast_gicp_st.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [CMakeFiles/fast_gicp.dir/src/fast_gicp/gicp/fast_vgicp.cpp.o] Error 1
make[2]: *** [CMakeFiles/fast_gicp.dir/src/fast_gicp/gicp/fast_gicp.cpp.o] Error 1
make[1]: *** [CMakeFiles/fast_gicp.dir/all] Error 2
make: *** [all] Error 2

Thank you if you can provide it.

**Comment by koide3** [comment]
I usually use ```libeigen3-dev``` installed via apt. What is the version of your OS?

**Comment by IamJasonYe** [comment]
I am using Ubuntu 16.04. And ROS is kinetic.

Currently, I solved my problem by using your provided dockerfile of melodic version. Also, I have tested the kinetic version of docker file, which has the same issue. One of them is about the CMake version, another one is about the PCL and eigen3 dependencies compile error.

**Comment by IamJasonYe** [comment]
So I guessed it's because both the ROS and Ubuntu 16.04 are too old and getting hard to getting the latest upgrade from the official repo.

**Comment by koide3** [comment]
Yes. I've tested this package on only Ubuntu 18 and 20, and maybe it cannot be built on 16.

**Comment by YangSiri** [comment]
Hi, @IamJasonYe  @koide3 

So the conclusion is that this repo can not be compatible with unbuntu 16.04 ?

**Comment by IamJasonYe** [comment]
> Hi, @IamJasonYe @koide3
> 
> So the conclusion is that this repo can not be compatible with unbuntu 16.04 ?

Yes, the current packages of ubuntu 16.04 via apt are obsolete. But if you can install those dependencies from source code, I guess you still can use it on 16.04, but which might be too tricky for quick reproduction or further deployment.

**Comment by YangSiri** [comment]
Hi, @IamJasonYe 

Okay, thank you. I change the deafault cmake version in ubuntu16 from 3.5 to 3.10 and rebuilt dependencies such as sophus and ceres with new version of eigen. However, it seems that there is still other errors when build __fast_gicp__.
